rls/rls-cargo-build-integration-notes.md

```diff
@@ -0,0 +1,105 @@
+
+Generating the build plan and its format are currently unstable and not yet used to integrate with other build systems or RLS.
+
+Note: When emitting build plan for `cargo check` is supported, the RLS could possibly drop a dependency the Cargo library.
+
+### What Cargo does (`cargo {build,check}`)
+An arbitrary breakdown that we think might be useful to identify separate, replaceable steps:
+1. Configuration (incl. package overrides etc.)
+2. Version resolution
+3. Package dependency graph construction (Cargo.lock)
+4. Action graph construction
+    1. Feature resolution - may change package deps based on features selected
+    2. Lowering dependencies to crate target level (proc macros, bins, libs)
+    3. Spliting nodes further into actions (e.g. compile build script, run build script, compile lib)
+    4. Differentiate/prune actions based on target dirtiness (fingerprints)
+5. Fetching packages (locate on disk, fetch from network)
+6. Build plan construction - flattened view into action graph mapped into actual system commands
+7. Execute the commands!
+
+# Possible plan
+## Split Cargo into 3 parts
+Having the above sequence of steps for `cargo build`, we can group these into:
+1. Front (`cargo build` -> action graph) (Step 1-5)
+2. Build orchestration (action graph -> build plan) (Step 6)
+3. Execution (invoking the build plan commands) (Step 7)
+
+With this we can mix and match:
+1. RLS should always execute the build (needs to invoke `rustc` itself to perform analysis)
+2. External build system can provide its set of packages and dependencies between them and Cargo can orchestrate that into a build plan (understandable by the RLS)
+3. Cargo can execute its front, fetch packages from crates.io, form an action graph, which can be further reused and translated into a build plan by the external build system
+4. etc.
+
+## Concrete goals
+### RLS <> other build systems
+* Establish a preliminary build plan format (probably use/adapt current Cargo `--build-plan`)
+* RLS needs to understand the build plan and be able to execute it in a sandbox (think `target/rls` dir), while consuming the save-analysis and diagnostics
+### RLS <> Cargo
+* Implement `cargo check ... --build-plan` in Cargo
+### RLS <> Buck
+* Buck needs to be able to tell the (almost) exact `rustc` command used to compile a crate target (embed the command in save-analysis?)
+* Translate Buck rules/generated save-analysis into build plan
+### Cargo <> other build systems
+* Establish a preliminary, Cargo action graph format - also notion of package version, targets (incl. build scripts and proc macros), Target/Host arch, Cargo features(?), declarative `rustc` args (profiles)
+* Allow Cargo to be fed an action graph format to be later executed by it
+* Emit also input files, in addition to command outputs, in the build plan to provide file-level dependencies
+* What more?
```

Comment:
I'll try and expand this list further tomorrow, meanwhile @nrc what do you suggest should also be included here, if anything?

Reply:
Later we might want to break up step 1 so other build systems can use bits of Cargo, but that is not in scope for the early stages of this work.